src/_P037_MQTTImport.ino

```diff
@@ -64,52 +76,118 @@ boolean Plugin_037(byte function, struct EventStruct *event, String& string)
+#endif
+#if (defined(P037_MAPPING_SUPPORT) || defined(P037_FILTER_SUPPORT)) && defined(P037_JSON_SUPPORT)
+                                          ,
+#endif
+#ifdef P037_JSON_SUPPORT
+                                          P037_PARSE_JSON
+#endif
+                                         );
+        delete P037_data;
         break;
       }
 
     case PLUGIN_WEBFORM_SAVE:
       {
-        String error;
-        char deviceTemplate[VARS_PER_TASK][41];		// variable for saving the subscription topics
-        for (byte varNr = 0; varNr < VARS_PER_TASK; varNr++)
-        {
-          String argName = F("p037_template");
-          argName += varNr + 1;
-          if (!safe_strncpy(deviceTemplate[varNr], web_server.arg(argName).c_str(), sizeof(deviceTemplate[varNr]))) {
-            error += getCustomTaskSettingsError(varNr);
-          }
-        }
-        if (error.length() > 0) {
-          addHtmlError(error);
+#ifdef P037_JSON_SUPPORT
+        P037_PARSE_JSON     = getFormItemInt(F("p037_parse_json"));
+#endif
+#ifdef P037_MAPPING_SUPPORT
+        P037_APPLY_MAPPINGS = getFormItemInt(F("p037_apply_mappings"));
+#endif
+#ifdef P037_FILTER_SUPPORT
+        P037_APPLY_FILTERS  = getFormItemInt(F("p037_apply_filters"));
+#endif
+        P037_SEND_EVENTS    = isFormItemChecked(F("p037_send_events"));
+
+        P037_data_struct *P037_data = new (std::nothrow) P037_data_struct(event->TaskIndex);
+
+        if (nullptr == P037_data) {
+          return success;
         }
-
-        SaveCustomTaskSettings(event->TaskIndex, (byte*)&deviceTemplate, sizeof(deviceTemplate));
-
-        success = true;
+        success = P037_data->webform_save(
+#ifdef P037_FILTER_SUPPORT
+                                          P037_APPLY_FILTERS
+#endif
+#if defined(P037_FILTER_SUPPORT) && defined(P037_JSON_SUPPORT)
+                                          ,
+#endif
+#ifdef P037_JSON_SUPPORT
+                                          P037_PARSE_JSON
+#endif
+                                         );
+        delete P037_data;
         break;
       }
 
     case PLUGIN_INIT:
       {
         success = false;
+        initPluginTaskData(event->TaskIndex, new (std::nothrow) P037_data_struct(event->TaskIndex));
+
         //    When we edit the subscription data from the webserver, the plugin is called again with init.
         //    In order to resubscribe we have to disconnect and reconnect in order to get rid of any obsolete subscriptions
         if (MQTTclient_connected) {
           //		Subscribe to ALL the topics from ALL instance of this import module
           MQTTSubscribe_037(event);
           success = true;
         }
+        break;
       }
-      break;
+
+    case PLUGIN_EXIT:
+      {
+        P037_data_struct *P037_data = static_cast<P037_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+        if (nullptr == P037_data) {
+          return success;
+        }
+
+        clearPluginTaskData(event->TaskIndex);
```

Comment:
Not sure if this is needed anymore, isn't that done after a call to `PLUGIN_EXIT` ?

Reply:
I'll test that and look at the __Plugin.cpp code to be sure.

Reply:
Removed that call, as it is done in `Plugin.cpp`